src/shred_collector/shred.zig

```diff
@@ -71,26 +76,58 @@ pub const Shred = union(ShredType) {
         }
     }
 
-    pub fn commonHeader(self: *const Self) *const ShredCommonHeader {
+    pub fn merkleRoot(self: Self) !Hash {
+        return switch (self) {
+            inline .Code, .Data => |s| getMerkleRoot(
+                s.fields.payload,
+                @TypeOf(s.fields).constants,
+                s.fields.common.shred_variant,
+            ),
+        };
+    }
+
+    pub fn chainedMerkleRoot(self: Self) !Hash {
+        return switch (self) {
+            inline .Code, .Data => |s| layout.getChainedMerkleRoot(s.fields.payload) orelse
+                error.InvalidPayloadSize,
+        };
+    }
+
+    pub fn isLastInSlot(self: *const Self) bool {
         return switch (self.*) {
-            inline .Code, .Data => |c| &c.fields.common,
+            .Code => false,
+            .Data => |data| data.fields.custom.flags.isSet(.last_shred_in_slot),
+        };
+    }
+
+    pub fn verify(self: Self, signer: sig.core.Pubkey) bool {
+        return switch (self) {
+            inline .Data, .Code => |s| s.fields.verify(signer),
         };
     }
 };
 
+/// Analogous to [ShredCode](https://github.com/anza-xyz/agave/blob/7a9317fe25621c211fe4ab5491b88a4757d4b6d4/ledger/src/shred/merkle.rs#L74)
 pub const CodingShred = struct {
     fields: Fields,
     const Fields = GenericShred(CodingShredHeader, coding_shred);
 
     const Self = @This();
     const consts = coding_shred;
 
-    fn sanitize(self: *const Self) error{InvalidNumCodingShreds}!void {
+    pub fn firstCodingIndex(self: *const Self) ?u32 {
```

Comment:
a bit out of scope but theres a method `first_coding_index` which returns an error instead of optional - probs should choose one or the other and use a `catch` or and `orelse`

Reply:
d29e2fd0edccdaee741be8b3cfceca78468eb542